Adds new features to Ion Data Generator to process constraints of Timestamp, Clob and Decimal.

linlin-s opened the pull request:
**Description of changes:**
In this pull request, more features were added to support Ion Data Generator to process more constraints provided by Ion Schema. The supported Ion Schema constraints in this PR includes:

- Timestamp:
```
   timestamp_precision:<RANGE<TIMESTAMP_PRECISION_VALUE>>
```
- Decimal:
```
   precision: <INT>
   precision: <RANGE<INT>>
   scale: <INT>
   scale: <RANGE<INT>>
 ```
- Clob:
 ```
   byte_length: <INT>
   byte_length: <RANGE<INT>>
 ```

1. IonSchedmaUtilities.java

- Adds logic to process timestamp_precision:<RANGE<TIMESTAMP_PRECISION_VALUE>> to the method `IonSchemaUtilities.parseConstraints`.

2. WriteRandomIonValues.java

- Adds method `WriteRandomIonValues.writeRandomDecimalsFromSchema` which is able to process the constraints of decimal.
- Adds method `WriteRandomIonValues.writeRandomTimestampsFromSchema` which is able to process the constraints of timestamp.
- Adds parameter constraintStruct to method `WriteRandomIonValues.writeRandomLobs` to allow the method to process the constraints of clob.
- Add method `WriteRandomIonValues.constructDecimalFromSchema` which is able to construct the decimal conforming with the constraints in Ion Schema.

**Test:**
- Adds two ISL file (testTimestamp.isl, testDecimal.isl) as inputs to test the Ion Data generator .
- Adds methods to test whether there is violation when generating timestamps , clobs and decimals.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

linlin-s committed:
Adds new features to Ion Data Generator to process constraints of Timestamp, Clob and Decimal.

linlin-s committed:
Adds unit test for generating Clob process.

desaikd commented on tst/com/amazon/ion/benchmark/DataGeneratorTest.java:
```diff
@@ -288,6 +291,33 @@ public void testViolationOfNestedIonStruct() throws Exception {
+     * @throws Exception if error occurs during the violation detecting process.
+     */
+    @Test
+    public void testViolationOfIonClob() throws Exception {
```
Should there be a test for Ion Blob with `byte_length` constraint?

linlin-s replied:
> 

Yes, the blob test will be added.

desaikd commented on src/com/amazon/ion/benchmark/WriteRandomIonValues.java:
```diff
@@ -52,7 +53,7 @@
     final static private List<Integer> NO_COEFFICIENT_DIGIT_RANGE = null;
     final static private String NO_TIMESTAMP_TEMPLATE = null;
-    final static private IonStruct NO_CONSTRAINT_STRUCT = null;
+    final static public IonStruct NO_CONSTRAINT_STRUCT = null;
```
Can we please add a comment around usage of `NO_CONSTRAINT_STRUCT` constant?

linlin-s replied:
> Can we please add a comment around usage of `NO_CONSTRAINT_STRUCT` constant?

Will add the general usage explanation for these constant. However, these constant will be considered removed because of the refactoring in the future PR.

desaikd commented on src/com/amazon/ion/benchmark/WriteRandomIonValues.java:
```diff
@@ -160,6 +161,23 @@ public static void writeRandomDecimals(int size, IonType type, String path, Stri
+     * @param constraintStruct is an IonStruct which contains the top-level constraints in Ion Schema.
+     * @throws Exception if an error occurs when building up the writer.
+     **/
+    public static void writeRandomDecimalsFromSchema(int size, IonType type, String path, String format, IonStruct constraintStruct) throws Exception {
```
Both `writeRandomDecimalsFromSchema` and `writeRandomTimestampsFromSchema` seem to have duplicate code. We can probably have a single method to reuse here.

linlin-s replied:
> Both `writeRandomDecimalsFromSchema` and `writeRandomTimestampsFromSchema` seem to have duplicate code. We can probably have a single method to reuse here.

Good catch, this would be considered as a part of refactoring and reconstruct the code. I'll create an issue on this question. Will be updated in the next PR.

desaikd commented on tst/com/amazon/ion/benchmark/DataGeneratorTest.java:
```diff
@@ -288,6 +291,33 @@ public void testViolationOfNestedIonStruct() throws Exception {
 
+    /**
+     * Test if there's violation when generating Ion Timestamp based on Ion Schema.
+     * @throws Exception  if error occurs during the violation detecting process.
```
```suggestion
     * @throws Exception if error occurs during the violation detecting process.
```

desaikd commented on src/com/amazon/ion/benchmark/WriteRandomIonValues.java:
```diff
@@ -381,13 +416,19 @@ public static Timestamp writeTimestamp(Timestamp.Precision precision,  Timestamp
      * @throws Exception if an error occurs when building up the writer.
     */
-    public static void writeRandomLobs(int size, IonType type, String format, String path) throws Exception {
+    public static void writeRandomLobs(int size, IonType type, String format, String path, IonStruct constraintStruct) throws Exception {
```
add a doc comment for parameter `constraintStruct`. Also, the doc comments for the parameters can be ordered as they appear in the function call. 
```
     /** 
     * ...
     * @param size specifies the size in bytes of the generated file.
     * @param type determines which type of data will be generated [blob | clob].
     * @param format ...
     * @param path ...
     * @param constraintStruct ...
     */
```

desaikd reviewed (commented):
Looks like a good start! I have left some minor comments for this code change.

tgregg commented on src/com/amazon/ion/benchmark/IonSchemaUtilities.java:
```diff
@@ -82,6 +87,8 @@ public static int parseConstraints(IonStruct value, String keyWord) throws IOExc
                                 if (reader.symbolValue().equals(KEYWORD_MAX)) {
                                     max = Integer.MAX_VALUE;
+                                } else if (keyWord.equals(IonSchemaUtilities.KEYWORD_TIMESTAMP_PRECISION)) {
+                                    max = Timestamp.Precision.valueOf(reader.stringValue().toUpperCase()).ordinal();
```
What if the first element is an int and the second element is a timestamp? Should that be an error?

linlin-s replied:
> What if the first element is an int and the second element is a timestamp? Should that be an error?

The value processed by this method doesn't contain the combinations of int and timestamp. Normally this method will process the value like [min/int, max/int] ,[min/timestamp.precision, max/timestamp.precision], int. Correct me if I misunderstand this question.

tgregg replied:
Well-formed Ion schema data won't have this problem. My question is: what happens if you receive an invalid schema? This question applies more broadly than just here. I want you to think about what kind of validation might be necessary. Or, is there a way to use an Ion Schema library to pre-validate the schema so that this doesn't need to be done in the data generator?

linlin-s replied:
> Well-formed Ion schema data won't have this problem. My question is: what happens if you receive an invalid schema? This question applies more broadly than just here. I want you to think about what kind of validation might be necessary. Or, is there a way to use an Ion Schema library to pre-validate the schema so that this doesn't need to be done in the data generator?

The method of checking validation of schema has been updated in the last commit. The method IonSchemaUtilities.checkValidationOfSchema will be called at the start place where ion schema file is loaded. This method is using the method `newSchema(inputSchema.iterator())` of ` ion schema kotlin` . If there's invalid schema syntax in the `inputSchema`, the `InvalidSchemaException` will be catched.

tgregg replied:
Nice!

tgregg commented on src/com/amazon/ion/benchmark/WriteRandomIonValues.java:
```diff
@@ -160,6 +161,23 @@ public static void writeRandomDecimals(int size, IonType type, String path, Stri
+    /**
+     * Write random Ion decimals into the file, and all generated data is conformed with the constraints provided by Ion schema file.
+     * @param size specifies the size in bytes of the generated file.
+     * @param type determines which type of ion data will be generated.
```
Is `type` implicit, since this method is for writing decimals? What happens if a non-decimal type is provided?

linlin-s replied:
> Is `type` implicit, since this method is for writing decimals? What happens if a non-decimal type is provided?

I think 'type' should be explicit parameter because it was predefined as 'DECIMAL' before calling this method. The reason passing the parameter 'type' is for the further usage of calling method  WriteRandomIonValues.writeRequestedSizeFile.

tgregg replied:
Is it always DECIMAL when calling this method?

linlin-s replied:
> Is it always DECIMAL when calling this method?

Yes, it is always DECIMAL

tgregg replied:
Then why does it need to be passed into this method? Can't this method just pass IonType.DECIMAL to the method it calls?

linlin-s replied:
> Then why does it need to be passed into this method? Can't this method just pass IonType.DECIMAL to the method it calls?

The change will be updated in the next PR which includes refactoring of these methods.

tgregg commented on src/com/amazon/ion/benchmark/IonSchemaUtilities.java:
```diff
@@ -32,8 +32,11 @@
+    public static final String KEYWORD_PRECISION = "precision";
 
     /**
      * Extract the value of the constraints, select from the set (occurs | container_length | codepoint_length).
```
Looks like this comment needs to be updated.

tgregg commented on src/com/amazon/ion/benchmark/WriteRandomIonValues.java:
```diff
@@ -160,6 +161,23 @@ public static void writeRandomDecimals(int size, IonType type, String path, Stri
+    public static void writeRandomDecimalsFromSchema(int size, IonType type, String path, String format, IonStruct constraintStruct) throws Exception {
+        File file = new File(path);
+        try (IonWriter writer = WriteRandomIonValues.formatWriter(format, file)) {
+            WriteRandomIonValues.writeRequestedSizeFile(size, writer, file, type, NO_CODE_POINT_LENGTH, NO_POINT_RANGE, NO_EXPONENT_VALUE_RANGE, NO_COEFFICIENT_DIGIT_RANGE, NO_TIMESTAMP_TEMPLATE, constraintStruct);
```
Seeing a method with this many parameters makes me think a refactoring is probably a good idea. We talked offline about deprecating the individual options for specifying constraints manually, in favor of doing all generation via schema. I think that could help with this.

linlin-s replied:
> Seeing a method with this many parameters makes me think a refactoring is probably a good idea. We talked offline about deprecating the individual options for specifying constraints manually, in favor of doing all generation via schema. I think that could help with this.

I'll create an issue on this. The next PR will include the refactoring.

tgregg commented on src/com/amazon/ion/benchmark/WriteRandomIonValues.java:
```diff
@@ -604,6 +653,23 @@ public static BigDecimal constructDecimal(List<Integer> expValRange, List<Intege
+        Random random = new Random();
+        StringBuilder rs = new StringBuilder();
+        for (int digit = 0; digit < precision; digit++) {
+            rs.append(random.nextInt(9) + 1);
```
Should this be `rs.append(random.nextInt(10));`?

linlin-s replied:
> Should this be `rs.append(random.nextInt(10));`?

This is used for generating random integer from 1 to 9 which is following this formula random.nextInt((max - min) + 1) + min

tgregg replied:
Right, but don't you want to generate random integers between 0 and 9 (except for the first digit, which should be between 1 and 9)?

linlin-s replied:
> Right, but don't you want to generate random integers between 0 and 9 (except for the first digit, which should be between 1 and 9)?

Right! Thank you for pointing out. The first digit should be random int between 1 to 9, and the rest of the digits should be random int between 0 to 9 . This will be updated in the following commit.

linlin-s committed:
Updates based on the comments
Added unit test for generating blobs process.
Added comments for unclear constant.
Updated doc comment for method WriteRandomIonValues.writeRandomLobs.
Updated doc comment for method IonSchemaUtilities.parseConstraints.
Updated the method WriteRandomIonValues.constructDecimalFromSchema.

linlin-s committed:
Adds method to check the validation of input ion schema file.
